Fix part of #10474: Make typescript checks strict for core/templates/components/graph-services/graph-layout.service.ts file

ayushjaink8 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #10474 .
2. This PR does the following: Make typescript checks strict for core/templates/components/graph-services/graph-layout.service.ts file.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
![image](https://user-images.githubusercontent.com/75239620/143188114-879ca1d5-0b26-4b79-a291-a6e66765686f.png)


## A Note for reviewers:
One front-end test is failing, so I have used --no-verify to push this code.
Error:
![image](https://user-images.githubusercontent.com/75239620/143193305-dcaf8ae7-dc74-4f0c-828c-40d4c2241da4.png)


<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

ayushjaink8 committed:
graph-layout-services.ts typescipt checks strict

ayushjaink8 committed:
commented file name inside tsconfig-strict.json

ayushjaink8 committed:
changes

ayushjaink8 committed:
alternative solution

ayushjaink8 committed:
removed commented file in tsconfig-strict.json

ayushjaink8 committed:
removed linting error

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -50,18 +50,18 @@ interface NodeData {
-  width: number;
-  id: string;
-  label: string;
+  y0: number | null;
+  x0: number | null;
+  yLabel: number | null;
+  xLabel: number | null;
+  height: number | null;
+  width: number | null;
+  id: string | null;
+  label: string | null;
```
Why can these be `null`?

ayushjaink8 replied:
![image](https://user-images.githubusercontent.com/75239620/143452078-1382622a-e962-4e89-bec4-ee6a9ddddec9.png)
here (in the same file: 240-255), as you can see we are declaring these as null.

vojtechjelinek replied:
Yeah, that is the explanation of usage, but not of the purpose, why are these set to `null` at that place? Is it possible to not use `null`? If not then it should be documented and explained what it means.

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -50,18 +50,18 @@ interface NodeData {
 }
 
-interface NodeDataDict {
+interface NodeDataDict{
```
Why this change?

ayushjaink8 replied:
Sorry, this was done by mistake... I will revert it in my next commit!

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -73,11 +73,11 @@ export class StateGraphLayoutService {
   // The last result of a call to computeLayout(). Used for determining the
   // order in which to specify states in rules.
-  lastComputedArrangement: NodeDataDict = null;
+  lastComputedArrangement: NodeDataDict | undefined;
```
Why this change? Why not
```suggestion
  lastComputedArrangement: NodeDataDict | null = null;
```

ayushjaink8 replied:
well I am not sure about this.... we can do that but then we need to add 
`getLastComputedArrangement(): NodeDataDict | null {`   - at line 488 
instead of  `getLastComputedArrangement(): NodeDataDict | undefined {` which I committed.

Can you please tell me the difference between undefined and null over here? I am not able to get this

vojtechjelinek replied:
They are basically the same, you can read some resources online. I think in our codebase we usually prefer to use `null` in cases where we return something that is logically empty. `undefined` usually means that something is missing (i.e. not defined).

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -485,7 +485,7 @@ export class StateGraphLayoutService {
   }
 
-  getLastComputedArrangement(): NodeDataDict {
+  getLastComputedArrangement(): NodeDataDict | undefined {
```
Why can this return `undefined`?

ayushjaink8 replied:
Because at line 76, I have committed this code:  `lastComputedArrangement: NodeDataDict | undefined;`. 
if we change `undefined` to `null` at line 76 then we can do the same at this line also.

Should I do it??

vojtechjelinek replied:
Yep, probably use `null` but add a comment explanation when this can be `null`.

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -500,15 +500,15 @@ export class StateGraphLayoutService {
     for (var nodeId in nodeData) {
       leftEdge = Math.min(
-        nodeData[nodeId].x0 - BORDER_PADDING, leftEdge);
+        Number(nodeData[nodeId].x0) - BORDER_PADDING, leftEdge);
```
Why this change?

ayushjaink8 replied:
because nodeData[nodeId].x0 can be null so it was showing a Ts error (as we can not then use it inside "min" function) so to deal with it I wrapped it inside Number(). I am not sure if it is the best way of dealing with such cases.
We can also use if-else over here. (like I have done at lines 551-552)

vojtechjelinek replied:
The best case is usually to get rid of the possibility of it being `null`, the second best thing to do is to use `if`.

vojtechjelinek replied:
ditto elsewhere

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -520,7 +520,8 @@ export class StateGraphLayoutService {
 
   getAugmentedLinks(
-      nodeData: NodeDataDict, nodeLinks: GraphLink[]): AugmentedLink[] {
+      nodeData: NodeDataDict,
+      nodeLinks: GraphLink[]): AugmentedLink[] | undefined {
```
```suggestion
      nodeData: NodeDataDict,
      nodeLinks: GraphLink[]
  ): AugmentedLink[] | undefined {
```

Why can it return `undefined`?

ayushjaink8 replied:
I did this to remove the TS error given below:
`core/templates/components/graph-services/graph-layout.service.ts(542,11): error TS2322: Type 'undefined' is not assignable to type 'AugmentedLink[]'.`

Refer lines 541-543 of the same file, you will see we are not returning anything.
![image](https://user-images.githubusercontent.com/75239620/143454765-d59b910e-9b42-4c88-a28a-8dc311705a2b.png)

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -547,64 +548,65 @@ export class StateGraphLayoutService {
-            (dx === 0) ? (targetHeight / 2) / Math.abs(dy) :
-            Math.min(endCutoff, (targetHeight / 2) / Math.abs(dy)));
-        }
+        if (targetx && sourcex && targety && sourcey && targetWidth &&
+          sourceWidth && targetHeight && sourceHeight) {
```
Why is this needed?

ayushjaink8 replied:
because all these values can be null in lines 55-58:
 ` yLabel: number | null;`
 ` xLabel: number | null;`
 ` height: number | null;`
 ` width: number | null;`

so as given at lines 537-540:
`  var sourcex = link.source.xLabel;`
`  var sourcey = link.source.yLabel;`
`  var targetx = link.target.xLabel;`
`  var targety = link.target.yLabel;`    
and at lines 547-550:
`  var sourceWidth = link.source.width;`
`  var sourceHeight = link.source.height;`
`  var targetWidth = link.target.width;`
`  var targetHeight = link.target.height;` 

these values (i.e, targetx, sourcex, targety, sourcey, targetWidth, sourceWidth, targetHeight, sourceHeight)
can also be `null`. Therefore to avoid  the following TS errors:

`core/templates/components/graph-services/graph-layout.service.ts(566,22): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(566,32): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(567,22): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(567,32): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(572,20): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(572,49): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(573,20): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(573,49): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(574,22): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(575,22): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(576,20): error TS2531: Object is possibly 'null'.`

I have written this if-else statement. We can also put them inside ` Number()` as done at line 503.

.

vojtechjelinek replied:
The best approach would again be to get rid of the `null`s, or at least make only one of the values indicate the the rest will be `null`. But if we keep the if it should look like this.
```suggestion
        if (
          target !== null && 
          sourcex !== null && 
          targety !== null && 
          sourcey !== null &&
          targetWidth  !== null && 
          sourceWidth !== null && 
          targetHeight !== null && 
          sourceHeight !== null 
        ) {
```

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -547,64 +548,65 @@ export class StateGraphLayoutService {
   modifyPositionValues(
-      nodeData: NodeDataDict, graphWidth: number,
-      graphHeight: number): NodeDataDict {
+      nodeData: Record<string, number>[], graphWidth: number,
+      graphHeight: number): Record<string, number>[] {
```
```suggestion
      nodeData: Record<string, number>[], 
      graphWidth: number,
      graphHeight: number
  ): Record<string, number>[] {
```

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -547,64 +548,65 @@ export class StateGraphLayoutService {
-        nodeData[nodeId][VERTICAL_NODE_PROPERTIES[i]] = (
-          graphHeight *
-          nodeData[nodeId][VERTICAL_NODE_PROPERTIES[i]]);
+        let nodeeData = nodeData[nodeId] as Record<string, number>;
```
Why is this needed?

ayushjaink8 replied:
To remove this TS error:
`core/templates/components/graph-services/graph-layout.service.ts(597,9): error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'NodeData'.
  No index signature with a parameter of type 'string' was found on type 'NodeData'.`
`core/templates/components/graph-services/graph-layout.service.ts(599,11): error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'NodeData'.
  No index signature with a parameter of type 'string' was found on type 'NodeData'.`
`core/templates/components/graph-services/graph-layout.service.ts(600,9): error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'NodeData'.
  No index signature with a parameter of type 'string' was found on type 'NodeData'.`
`core/templates/components/graph-services/graph-layout.service.ts(602,11): error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'NodeData'.
  No index signature with a parameter of type 'string' was found on type 'NodeData'.`

Actually, for this particular error, I took help from @mridul-netizen and he suggested me this code. The problem over here is we can not actually use string like `nodeData[nodeId]["string"]` will throw TS error. That's why we reinitialized `nodeData[nodeId]` as `Record<string, number>`.

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments, can you also strictly type the spec file of this service?

ayushjaink8 committed:
formatted modifyPositionValues in core/templates/components/graph-services/graph-layout.service.ts

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

ayushjaink8 committed:
Update core/templates/components/graph-services/graph-layout.service.ts

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

ayushjaink8 committed:
Merge branch 'oppia:develop' into ayushjain

ayushjaink8 committed:
added typescript check strict for graph-layout-service.spec.ts

ayushjaink8 committed:
fixed merge conflicts

ayushjaink8 commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -47,17 +47,18 @@ interface AugmentedLink {
 }
 
 interface NodeData {
+  [index: string]: number | boolean | string | null;
```
I added this line and removed that **Record<string, number>[]** from the `modifyPositionValues` (Refer line no. 588 to 613)
**which was done in the previous commit (165e258).**

I got the idea of doing these index signatures from here: https://www.typescriptlang.org/docs/handbook/2/objects.html#index-signatures](url)

![image](https://user-images.githubusercontent.com/75239620/143617988-22fa22d1-4ed8-43b4-b3d2-eaaa56326977.png)

ayushjaink8 replied:
@vojtechjelinek  are you able to see the above comment at your end??

vojtechjelinek replied:
Wait I'm not sure I understand why you did this.

ayushjaink8 committed:
removed lint errors

ayushjaink8 commented:
> Thanks! Left a few comments, can you also strictly type the spec file of this service?

Done! waiting for your response

ayushjaink8 committed:
Merge branch 'oppia:develop' into ayushjain

ayushjaink8 commented:
@vojtechjelinek Can you please have a look at my PR and give your review?

PC: I had replied to all of your comments.

ayushjaink8 committed:
fixed lint errors

ayushjaink8 committed:
Merge branch 'ayushjain' of https://github.com/ayushjaink8/oppia into ayushjain

vojtechjelinek commented:
> @vojtechjelinek Can you please have a look at my PR and give your review?
> 
> PC: I had replied to all of your comments.

Hey, I don't see any replies to my comments. 

![image](https://user-images.githubusercontent.com/10142938/143780782-660e65c6-6d00-4592-bddc-d9610f8637b9.png)

ayushjaink8 commented:
> > @vojtechjelinek Can you please have a look at my PR and give your review?
> > PC: I had replied to all of your comments.
> 
> Hey, I don't see any replies to my comments.
> 
> ![image](https://user-images.githubusercontent.com/10142938/143780782-660e65c6-6d00-4592-bddc-d9610f8637b9.png)

hi @vojtechjelinek 
**I don't know how is this possible but it is showing all the comments at my end. I forgot to write @vojtechjelinek  in those comments. Can it be because of that reason?**

![image](https://user-images.githubusercontent.com/75239620/143804715-6893f431-4dd5-4985-88cc-8eed16664c25.png)
__________________________________________________________________________________________________
![image](https://user-images.githubusercontent.com/75239620/143804832-fb1dad88-6166-4c46-8df3-654194f6dd88.png)
________________________________________________________________________________________________
![image](https://user-images.githubusercontent.com/75239620/143804985-092e6699-91e2-4a73-996b-bc55fccab71c.png)
_________________________________________________________________________________________________
![image](https://user-images.githubusercontent.com/75239620/143805140-3dff5f79-ae34-45db-ba40-d8804c208f88.png)
_________________________________________________________________________________________________


**I have added this comment on my own afterwards and seek your review, I have changed the previous code of `Record<string, number>` a little bit because it was showing an error when I tried to make the spec file type checks strict:**

![image](https://user-images.githubusercontent.com/75239620/143805234-d10f67df-6764-469e-b9f9-f971a026e998.png)


___________________________________________________________________________________

vojtechjelinek commented:
@ayushjaink8 The comments are pending. You need to submit them.

ayushjaink8 commented:
> @ayushjaink8 The comments are pending. You need to submit them.

@vojtechjelinek I have submitted the review, Are you able to see all my comments now??

ayushjaink8 committed:
Merge branch 'oppia:develop' into ayushjain

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -520,7 +521,9 @@ export class StateGraphLayoutService {
-      nodeData: NodeDataDict, nodeLinks: GraphLink[]): AugmentedLink[] {
+      nodeData: NodeDataDict,
+      nodeLinks: GraphLink[]
+  ): AugmentedLink[] | undefined {
```
Why can this return `undefined`?

ayushjaink8 replied:
@vojtechjelinek  Here, we are not returning anything. **Refer lines: 544-547**

![image](https://user-images.githubusercontent.com/75239620/144302490-58430bc9-fbdd-4f2e-8092-30b21f8168d0.png)

Also, **Lines: 485-527** inside **.spec.ts** file also have something to do with this. We are expecting it to be undefined when 
`State1.xLabel === State2.xLabel` and `State1.yLabel === State2.yLabel`
![image](https://user-images.githubusercontent.com/75239620/144303318-11477eeb-80e5-4237-853e-cf7eed80f1c0.png)

vojtechjelinek replied:
Okay, please add an explanatory code comment.

ayushjaink8 replied:
Done.

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -596,10 +604,10 @@ export class StateGraphLayoutService {
         nodeData[nodeId][HORIZONTAL_NODE_PROPERTIES[i]] = (
           graphWidth *
-          nodeData[nodeId][HORIZONTAL_NODE_PROPERTIES[i]]);
+          Number(nodeData[nodeId][HORIZONTAL_NODE_PROPERTIES[i]]));
```
Again, why `Number`.

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

ayushjaink8 committed:
Update core/templates/components/graph-services/graph-layout.service.ts

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

ayushjaink8 committed:
removed null from NodeDataDict

ayushjaink8 committed:
removed linting errors

ayushjaink8 commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -242,20 +242,20 @@ export class StateGraphLayoutService {
-        height: null,
-        width: null,
-        reachableFromEnd: false
+        x0: 0,
+        y0: 0,
+        xLabel: 0,
+        yLabel: 0,
+        id: '',
+        label: '',
+        height: 0,
+        width: 0,
+        reachableFromEnd: false,
```
@vojtechjelinek  It doesn't really matter what we initialize here for these values since we are going to change them later on in BFS. **Refer Lines: 439 - 457**
![image](https://user-images.githubusercontent.com/75239620/144301609-5c8d48b7-620c-4191-a0a9-565720849062.png)

These values are never going to be null according to my understanding so I tried to avoid `null` as much as I can.

vojtechjelinek replied:
It always matters. It might not matter now, but later some changes to the code might influence it. Ther preferred thing to do here is to get rid of filling the empty values. If that is not possible then we should use `null`s.

ayushjaink8 replied:
> 

@vojtechjelinek But when I add `null` then we need to add a lot of if-else statements at various places in order to remove the TS errors which again can cause various different errors later on. Since these are never going to be `null`s so I don't want to add ` number | null` over here

![image](https://user-images.githubusercontent.com/75239620/144884484-a2b0bcec-8c40-4e19-acac-0e8b0001ce79.png)

Is there any alternative to this situation because `null` doesn't seem to be a good option over here according to me. 
And as I have explained in the previous comment we are going to initialize these values according to the given function parameters later on. Also, these values represent pixel positions so `0` is the best initialization we should do..

Although, We can use `string | null` for `id` and `label`.

vojtechjelinek replied:
Why do we need to wait for later? Can we assign the correct values when we create the object?

ayushjaink8 replied:
@vojtechjelinek 

> Why do we need to wait for later? Can we assign the correct values when we create the object?

No, actually we are doing some calculations on the input values and the other values of `nodeData` like `depth: ` and `offset: `, and then we are assigning them. We are not directly assigning them. If you see the whole function then the first place where we are using these `y0, x0, yLabel, xLabel, height, width` is here:

![image](https://user-images.githubusercontent.com/75239620/145208519-a1e2070e-9144-4edb-a18f-412bdbf834c6.png)

Here, we are initializing these values along with the `id` and `label`.

vojtechjelinek replied:
But then we should use `null`, or you can somehow transform this so that you save these values into a different object and then create this object here.

ayushjaink8 replied:
@vojtechjelinek yeah! But still, even if we create a different object it will be of interface `NodeData` and we will get a TS error because we can not assign `null` to a `number` and if I change `number`  to `number | null` in the interface itself then a hell lot of if-else statements will be required at various places to remove other TS error.

ayushjaink8 replied:
@vojtechjelinek I am looking for a hack not to change the `interface NodeData` and add `null` there.

vojtechjelinek replied:
You can always create a new interface too.

ayushjaink8 replied:
@vojtechjelinek  I have done that, Please check my new changes!

ayushjaink8 commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -485,7 +485,7 @@ export class StateGraphLayoutService {
   }
 
-  getLastComputedArrangement(): NodeDataDict {
+  getLastComputedArrangement(): NodeDataDict | null {
     return cloneDeep(this.lastComputedArrangement);
```
This is `null` because we are initializing `lastComputedArrangement` also as `null`. 
**Refer Line: 76**

ayushjaink8 replied:
Also, inside **.spec.ts** file we are expecting it to be `null` before running `computeLayout()` function.
**Refer Line: 764** 

![image](https://user-images.githubusercontent.com/75239620/144308348-d9677e8c-7e8b-4422-ad27-ff362180594b.png)

vojtechjelinek replied:
Looks good. Please add an explanatory comment.

ayushjaink8 replied:
Added one. Kindly review it

ayushjaink8 committed:
Merge branch 'oppia:develop' into ayushjain

ayushjaink8 commented:
@vojtechjelinek  Can you please review my PR? Any more explanations/changes are most welcomed.

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -73,11 +73,11 @@ export class StateGraphLayoutService {
   // The last result of a call to computeLayout(). Used for determining the
   // order in which to specify states in rules.
-  lastComputedArrangement: NodeDataDict = null;
+  lastComputedArrangement: NodeDataDict | null = null;
```
Change the comment to include what does `null` mean.

vojtechjelinek replied:
Not done.

ayushjaink8 replied:
check now, actually, I forgot to commit that

vojtechjelinek replied:
Looks good.

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -504,11 +504,13 @@ export class StateGraphLayoutService {
       rightEdge = Math.max(
-        nodeData[nodeId].x0 + BORDER_PADDING + nodeData[nodeId].width,
-        rightEdge);
+        nodeData[nodeId].x0 +
+        BORDER_PADDING +
+        nodeData[nodeId].width, rightEdge);
```
Why is this changed?

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -504,11 +504,13 @@ export class StateGraphLayoutService {
       bottomEdge = Math.max(
-        nodeData[nodeId].y0 + BORDER_PADDING + nodeData[nodeId].height,
-        bottomEdge);
+        nodeData[nodeId].y0 +
+        BORDER_PADDING +
+        nodeData[nodeId].height, bottomEdge);
```
Why is this changed?

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -586,10 +590,14 @@ export class StateGraphLayoutService {
+      graphWidth: number,
+      graphHeight: number
+  ): NodeDataDict {
+    var HORIZONTAL_NODE_PROPERTIES: ('x0' | 'width' | 'xLabel')[] =
+      ['x0', 'width', 'xLabel'];
+    var VERTICAL_NODE_PROPERTIES: ('y0' | 'height' | 'yLabel')[] =
+      ['y0', 'height', 'yLabel'];
```
```suggestion
    var HORIZONTAL_NODE_PROPERTIES: ('x0' | 'width' | 'xLabel')[] = [
      'x0', 'width', 'xLabel'
    ];
    var VERTICAL_NODE_PROPERTIES: ('y0' | 'height' | 'yLabel')[] = [
      'y0', 'height', 'yLabel'
    ];
```

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

ayushjaink8 committed:
Merge branch 'oppia:develop' into ayushjain

ayushjaink8 committed:
Update core/templates/components/graph-services/graph-layout.service.ts

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

ayushjaink8 committed:
fixed few lint errors

ayushjaink8 committed:
added two comments

ayushjaink8 committed:
removed trailing spaces

ayushjaink8 committed:
changed strict to noImplicitUseStrict

ayushjaink8 commented:
@vojtechjelinek Did some changes and added asked comments. Kindly review it. 
Also, seek your guidance for that `number | null` changes inside my code. 
**Direct discussion link:** [https://github.com/oppia/oppia/pull/14311#discussion_r763183876](url)

ayushjaink8 commented:
@vojtechjelinek Use this link: https://github.com/oppia/oppia/pull/14311/files#r763183876

ayushjaink8 committed:
Merge branch 'oppia:develop' into ayushjain

ayushjaink8 committed:
created new nodeData_temp

ayushjaink8 committed:
changed nodeData_temp to nodeDataTemp

ayushjaink8 committed:
removed lint errors

ayushjaink8 committed:
removed lint errors

ayushjaink8 commented:
@vojtechjelinek I have created a new object `nodeDataTemp` which temporarily be treated as `nodeData` and I am going to do all the calculations on `nodeDataTemp` and later on, I am initializing `nodeData` using if-else statements to deal with `null` values as prescribed by you.
Kindly review and merge my PR. Any more suggestions and changes are always welcomed.

ayushjaink8 commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -436,30 +456,66 @@ export class StateGraphLayoutService {
       WIDTH_OFFSET_CHECKPOINT_ICON;
     }
 
-    // Assign id and label to each node.
-    for (var nodeId in nodeData) {
-      nodeData[nodeId].id = nodeId;
-      nodeData[nodeId].label = nodes[nodeId];
```
@vojtechjelinek I have assigned these `id` and `label` at the time of initialization itself. 
**Refer lines: 267-268**

ayushjaink8 committed:
comment added

vojtechjelinek reviewed (approved):
LGTM!

ayushjaink8 committed:
removed lint errors

ayushjaink8 commented:
> LGTM!

So now what? What I am supposed to do?

ayushjaink8 committed:
corrected typo error

ayushjaink8 committed:
Merge branch 'oppia:develop' into ayushjain

ayushjaink8 committed:
Merge branch 'oppia:develop' into ayushjain

mridul-netizen commented:
@ayushjaink8 PR seems to be LGTM but there are some e2e failures. Can you try digging up the reason for these failures ? Try looking at previous commits/logs and see what changes made them fail!

ayushjaink8 commented:
> @ayushjaink8 PR seems to be LGTM but there are some e2e failures. Can you try digging up the reason for these failures ? Try looking at previous commits/logs and see what changes made them fail!

@mridul-netizen  Yes! It is because of the new interface which I just created `nodeDataTemp`. I guess these tests expect some changes in `nodeData` but since now every calculation is being done on `nodeDataTemp` instead of `nodeData`, it is showing errors. In the end, I am just converting `nodeDataTemp` to `nodeData` to return `NodeDataDict` type value, while dealing with the `null` values using if-else statements at the same time.

**Refer before and after this commit:  9d06027**

ayushjaink8 committed:
Merge branch 'oppia:develop' into ayushjain

vojtechjelinek commented:
> > @ayushjaink8 PR seems to be LGTM but there are some e2e failures. Can you try digging up the reason for these failures ? Try looking at previous commits/logs and see what changes made them fail!
> 
> @mridul-netizen Yes! It is because of the new interface which I just created `nodeDataTemp`. I guess these tests expect some changes in `nodeData` but since now every calculation is being done on `nodeDataTemp` instead of `nodeData`, it is showing errors. In the end, I am just converting `nodeDataTemp` to `nodeData` to return `NodeDataDict` type value, while dealing with the `null` values using if-else statements at the same time.
> 
> **Refer before and after this commit: [9d06027](https://github.com/oppia/oppia/commit/9d06027f62e168e6c10001509f33e34327562a4d)**

E2e tests usually don't expect any changes in variables since e2e tests work like automated manual testing (e.g. click this button, fill this field; is some value in HTML changed). You will need to debug the tests and fix the code.

ayushjaink8 committed:
Merge branch 'oppia:develop' into ayushjain

ayushjaink8 committed:
debugging e2e and front end errors

vojtechjelinek reviewed (changes requested):
Marking as requested changes until the tests are properly resolved.

ayushjaink8 commented:
> Marking as requested changes until the tests are properly resolved.

Ohk @vojtechjelinek but as you can see when I commented the new `NodeDataTemp` from the code... all the errors are gone. It is due to the introduction of a new interface for sure. But how is that possible, I mean I am not changing the code and just replacing `nodeData` with `nodeDataTemp`. Can you please guide me, what could be the possible error over here?

ayushjaink8 committed:
Merge branch 'oppia:develop' into ayushjain

vojtechjelinek commented:
> > Marking as requested changes until the tests are properly resolved.
> 
> Ohk @vojtechjelinek but as you can see when I commented the new `NodeDataTemp` from the code... all the errors are gone. It is due to the introduction of a new interface for sure. But how is that possible, I mean I am not changing the code and just replacing `nodeData` with `nodeDataTemp`. Can you please guide me, what could be the possible error over here?

Not sure, I don't have time to debug the code. You should try to debug it yourself and see what is the exact issue. There are some debugging guides on the wiki.

ayushjaink8 committed:
debug e2e errors

ayushjaink8 committed:
Merge branch 'ayushjain' of https://github.com/ayushjaink8/oppia into ayushjain

ayushjaink8 committed:
Merge branch 'oppia:develop' into ayushjain

ayushjaink8 committed:
Merge branch 'ayushjain' of https://github.com/ayushjaink8/oppia into ayushjain

ayushjaink8 committed:
debug errors

ayushjaink8 committed:
reverted few linting issues

ayushjaink8 committed:
edited initialization of nodeData

ayushjaink8 committed:
final changes

ayushjaink8 committed:
strict to noImplicitUseStrict

ayushjaink8 commented:
@vojtechjelinek  All checks have passed. All the errors have been resolved. Kindly review it. @mridul-netizen also suggested me to modify a few lines of this code based upon his PR: https://github.com/oppia/oppia/pull/14468  (modifyPositionValues)
 
Kindly review these changes and merge my PR.

ayushjaink8 committed:
Merge branch 'oppia:develop' into ayushjain

vojtechjelinek commented on core/templates/components/graph-services/graph-layout.service.ts:
```diff
@@ -287,7 +280,7 @@ export class StateGraphLayoutService {
         // depth to process now, we ignore it for now and stick it back in
         // the queue to be processed later.
         if (bestPath.indexOf(linkTarget) !== -1 &&
-            nodeData[linkTarget].depth !== nodeData[currNodeId].depth + 1) {
+          nodeData[linkTarget].depth !== nodeData[currNodeId].depth + 1) {
```
```typescript
if (
  bestPath.indexOf(linkTarget) !== -1 &&
  nodeData[linkTarget].depth !== nodeData[currNodeId].depth + 1
) {
```

ayushjaink8 replied:
done!
PS: did similar indentations at line no. 88 and 289 too.

vojtechjelinek reviewed (changes requested):
Thanks! Just one small nit.

ayushjaink8 committed:
few identation changes

ayushjaink8 committed:
Merge branch 'ayushjain' of https://github.com/ayushjaink8/oppia into ayushjain

vojtechjelinek reviewed (approved):
LGTM!
